lib/html5ever/precompiled.ex

```diff
@@ -0,0 +1,232 @@
+defmodule Html5ever.Precompiled do
+  @moduledoc false
+
+  require Logger
+
+  @mix_target Mix.target()
+  @available_targets ~w(aarch64-apple-darwin x86_64-apple-darwin x86_64-unknown-linux-gnu x86_64-unknown-linux-musl arm-unknown-linux-gnueabihf arm-unknown-linux-musleabihf)
+  @available_nif_versions ~w(2.14 2.15 2.16)
+
+  @doc """
+  Returns the target triple for download or compile and load.
+
+  This function is translating and adding more info to the system
+  architecture returned by Elixir/Erlang to one used by Rust.
+
+  The returning string format is the following:
+
+    "nif-NIF_VERSION-ARCHITECTURE-VENDOR-OS-ABI"
+
+  ## Examples
+
+      iex> target()
+      {:ok, "nif-2.16-x86_64-unknown-linux-gnu"} 
+
+      iex> target()
+      {:ok, "nif-2.15-aarch64-apple-darwin"}
+
+  """
+  def target(config \\ target_config()) do
+    sys_arch = String.split(config.system_arch, "-")
+
+    case config.os_type do
+      {:unix, osname} ->
+        arch_os = normalize_arch_os(osname, sys_arch) |> Enum.join("-")
+
+        cond do
+          arch_os not in @available_targets ->
+            {:error, "precompiled NIF is not available for this target: #{inspect(arch_os)}"}
+
+          config.nif_version not in @available_nif_versions ->
+            {:error,
+             "precompiled NIF is not available for this NIF version: #{inspect(config.nif_version)}"}
+
+          true ->
+            {:ok, "nif-#{config.nif_version}-#{arch_os}"}
+        end
+
+      {:win32, _} ->
+        {:error, "precompiled NIFs for Windows are not supported right now"}
+    end
+  end
+
+  defp target_config do
+    system_arch = :erlang.system_info(:system_architecture) |> List.to_string()
+
+    system_arch =
+      case @mix_target do
+        :host ->
+          system_arch
+
+        :x86_64 ->
+          system_arch
+
+        # Targets supported by the Nerves project.
+        # See: https://hexdocs.pm/nerves/targets.html
+        mix_target when mix_target in [:rpi, :rpi0, :rpi1, :rpi2, :rpi3, :rpi4, :bbb, :osd32mp1] ->
```

Comment:
I don't think we should use Mix target. The Mix target can be named anything, those names are only conveniences. Instead you should pattern match on `system_arch` instead, like we do here: https://github.com/phoenixframework/esbuild/blob/0272743c3e1d5e4898a2e24c20276b4cf0370180/lib/esbuild.ex#L220-L242

Reply:
The target here is only for use cases like Nerves. They set the `MIX_TARGET` env var to configure the "cross compilation" to Nerves. If I don't do this, then my host sets the `system_arch` and we download the wrong NIF. I will double check if this is needed.

Reply:
Hrm... let's check with the nerves team. It may be other env variables to check.

Reply:
I managed to make this work with the help of Frank :)
It's now using the `TARGET_*` env vars.